# Python Journey

Introduction to python(BASIC)

In [ ]:
12

12

In [ ]:
a=10

In [ ]:
a=10
print(a)

10


In [ ]:
a=10; b=20; c=30
print(a,b,c)

10 20 30


In [ ]:
a="Anurag";b=23;d=42.34
print(a)
print(b)
print(c)

Anurag
23
30


In [ ]:
a,b,c=1,2,3
print(a,b,c)

1 2 3


In [ ]:
a=100000000
print(a)
a=10_00_00_000
print(a)
a=100_000_000
print(a)

100000000
100000000
100000000


In [ ]:
a=10
print(id(a))

11654664


In [ ]:
a=10
print(type(a))
b="hello"
print(type(b))
c=2+3j
print(a,b,c,type(c),id(c))

<class 'int'>
<class 'str'>
10 hello (2+3j) <class 'complex'> 140504245801072


In [ ]:
a=10
print(a,id(a))  #same id for a and b
b=10
print(b,id(b))

10 11654664
10 11654664


In [ ]:
firstideinint=10
print(a)

firstIdInt=10  # camel case
print(a)

FirstIdInt=10   # pascal case
print(a)

first_Id_Int=10   # snake case
print(a)


10
10
10
10


In [ ]:
v1a23r1=10
print(v1a23r1)

10


In [ ]:
#identifier cannot start with int error

1var=10     #error

In [ ]:
# spaces are not alowwed in identifiers

first var=10   #error

In [ ]:
#no special symbols are allowed to use
#with identifiers, other than

first$var=10   #error

In [ ]:
#keywords cannot be used as identifiers

import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


In [ ]:
# keywords cannot use as a identifiers

break=10  #error

In [ ]:
a=10
a="niet"

In [ ]:
a="text"
print(a)
a='text'
print(a)

a=''' line one
       line two'''
print(a)
a=""" line one
      line two
      line three"""
print(a)

text
text
 line one
       line two
 line one
      line two
      line three


In [ ]:
a='line \
 one'
print(a)

line  one


In [ ]:
a=10
print(a,type(a))

10 <class 'int'>


In [ ]:
a='23'
print(type(a))
b=int(a)
print(b,type(b))

<class 'str'>
23 <class 'int'>


In [ ]:
a=10
print(a,type(a))
a=2+3j
print(a,type(a))
a=12.34
print(a,type(a))
a=.3457
print(a)

10 <class 'int'>
(2+3j) <class 'complex'>
12.34 <class 'float'>
0.3457


# List

In [ ]:
l=[]
print(l,type(l))
k=list()
print(k,type(k))

[] <class 'list'>
[] <class 'list'>


#tuple

In [ ]:
t=(1,2,3.56)
print(t,type(t))
t=tuple()
print(t)

(1, 2, 3.56) <class 'tuple'>
()


In [ ]:
k=range(2)  #start,stop,step
print(k)

range(0, 2)


# Dictonary

In [ ]:
d={}
print(d,type(d))
d=dict()
print(d)
d={1:'one', 2:'two', 3:'three'}
print(d,type(d))

{} <class 'dict'>
{}
{1: 'one', 2: 'two', 3: 'three'} <class 'dict'>


# set

In [ ]:
s={}
print(s,type(s))
s={23}
print(s,type(s))
s=set()
print(s,type(s))

{} <class 'dict'>
{23} <class 'set'>
set() <class 'set'>


In [ ]:
a=True
print(a,type(a))
a=False
print(a,type(a))
a=bool(-1)
print(a)

True <class 'bool'>
False <class 'bool'>
True


In [ ]:
#compile time i/p

a=10
a=b=c=30
print(a,b,c)


30 30 30


In [ ]:
#runtime input

a=input("enter the number ")
print(a)

enter the number 12
12


In [ ]:
# return type input is str

a=input("Enter the data: ")
print(a,type(a))

Enter the data: 741
741 <class 'str'>


In [ ]:
# return type input is integer

a=int(input("Enter the data:"))
print(a,type(a))

Enter the data:125
125 <class 'int'>


In [ ]:
# return type input is float

a=float(input("Enter float data :"))
print(a,type(a))

Enter float data :23
23.0 <class 'float'>


In [ ]:
# return type input is complex

a= complex(input("Enter float data :"))
print(a,type(a))

Enter float data :471
(471+0j) <class 'complex'>


In [ ]:
# return type input is boolean

a=bool(input("enter the data "))
print(a)

enter the data 52
True


In [ ]:
# multiple input at a time

a,b,c=int(input()),int(input()),int(input())
print(a,b,c)

10
20
30
10 20 30


# Split function

In [ ]:
a="hello hi how are you? bye".split()
print(a)
a="hello hi how are you? bye".split(sep='h')
print(a)

['hello', 'hi', 'how', 'are', 'you?', 'bye']
['', 'ello ', 'i ', 'ow are you? bye']


In [ ]:
a="hello hi how are you? bye"
b=a.split(sep=' ',maxsplit=2)
print(b)

['hello', 'hi', 'how are you? bye']


In [ ]:
print("hello this is Anurag".split())

['hello', 'this', 'is', 'Anurag']


In [ ]:
s="hello this is Anurag"
print(s.split())

['hello', 'this', 'is', 'Anurag']


In [ ]:
s="hello this is Anurag"
print(s.split(sep='h'))

['', 'ello t', 'is is Anurag']


In [ ]:
s="hello this is Anurag"
print(s.split(sep='a'))

['hello this is Anur', 'g']


In [ ]:
s="hello this is Anurag"
print(s.split(sep=' ',maxsplit=1))
print(s.split(sep=' ',maxsplit=2))
print(s.split(sep=' ',maxsplit=3))

['hello', 'this is Anurag']
['hello', 'this', 'is Anurag']
['hello', 'this', 'is', 'Anurag']


In [ ]:
a="10 20 304 40 12 4 566".split()
print(a)

['10', '20', '304', '40', '12', '4', '566']


# Multiple inputs in a single line can be initialized with the help of split()

In [ ]:
# i/p => 10 20 30
# i/p => how are you?
# i/p => hello ravi how are you? => Error

a,b,c = input().split() #a,b,c = 10 20 30
print(a,b,c)
a,b,c = input().split() #a,b,c = how are you
print(a,b,c)

10 20 30
10 20 30
how are you
how are you


In [ ]:
#i/p
a,b,c = input().split(sep=',')#a,b,c = 10,20,30
print(a,b,c)

10,20,30
10 20 30


In [ ]:
a,b,c = input().split() # a,b,c=10 20 30=>return type str
print(a,type(a),) #a+6 => str+int =>error
print(b,type(b),a+"hi") # str1+str2=>str1str2
print(c,type(c))

10 20 30
10 <class 'str'>
20 <class 'str'> 10hi
30 <class 'str'>


# Map

In [ ]:
a,b,c = map(int,input().split())
print(a,type(a),a+34)
print(b,type(b))
print(c,type(c))

10 20 30
10 <class 'int'> 44
20 <class 'int'>
30 <class 'int'>


In [ ]:
a,b,c = map(complex,input().split())
print(a,type(a),a+34)
print(b,type(b))
print(c,type(c))

10 2+3j 13
(10+0j) <class 'complex'> (44+0j)
(2+3j) <class 'complex'>
(13+0j) <class 'complex'>


# Formated Strings

In [ ]:
# Q - Test Case
# Input: 10 20 30
# Output:
# a:10 ,type:< class int >
# b:20, type:< class int >
# c:30 ,type:< class int >

a, b, c = map(int, input().split())

print(f"a:{a} ,type:{type(a)}")
print(f"b:{b}, type:{type(b)}")
print(f"c:{c} ,type:{type(c)}")


10 20 30
a:10 ,type:<class 'int'>
b:20, type:<class 'int'>
c:30 ,type:<class 'int'>


In [ ]:
#  print(objects,sep=' ',end='\n')
# \n new line escape character\
# \t tab space escape character
# \b back space escape character...

print("Hello\nWorld")
print("Hello\tWorld")
print("Helloo\b!")


Hello
World
Hello	World
Helloo!


In [ ]:
a,b,c=map(int,input().split())
print(a,b,c,sep='->')
print(a,b,c,sep=' :) ')
print(a,b,c,sep='@@$')

12 354 567
12->354->567
12 :) 354 :) 567
12@@$354@@$567


In [ ]:
a,b,c=map(int,input().split())
print(a,b,c,sep='->',end='.')
print("line two",end='&')
print("line three",end=".")
print("line four")


12 3 56
12->3->56.line two&line three. 
line four


In [ ]:
# multiple line comment
'''
input  =>  1 2 3 4
output =>  a : 1 ,b : 2 , c : 3 , d : 4
'''
a,b,c,d=map(int,input().split())
print("a :",a,',b :',b,", c :",c,", d :",d)

print(f'text : {a}=>id',end='\n\n')
print(f' a:{a}, b:{b}, c:{c}, d:{d}')

1 2 3 4
a : 1 ,b : 2 , c : 3 , d : 4
text : 1=>id
 a:1, b:2, c:3, d:4


# **Operators**

#1.   Arithmetic<br>
    / (True division) ,+ , -, \* , // (floor division) ,% (modulo) ,** power<br>



---
Operator Precedence => Associativity Rule <br>
1) () L-to-R          <br>
2) ** R-TO-L            <br>
3) * , / , % , // L-to-R <br>
4) + , -  L-to-R        <br>








In [ ]:
a,b=map(int,input().split())
print(a,"+",b,'=',a+b)
print(f' {a} + {b} = {a+b}')
print(f' {a} - {b} = {a-b}')
print(f' {a} * {b} = {a*b}')
print(f' {a} / {b} = {a/b}')
print(f' {a} // {b} = {a//b}')
print(f' {a} % {b} = {a%b}')
print(f' {a} ** {b} = {a**b}')

13 3
13 + 3 = 16
 13 + 3 = 16
 13 - 3 = 10
 13 * 3 = 39
 13 / 3 = 4.333333333333333
 13 // 3 = 4
 13 % 3 = 1
 13 ** 3 = 2197


In [ ]:
a,b=-10,3
print(a/b)  #-3.3333333333333335

print(a//b) # roundups to lowest value
            # -4

print(a%b)  # a%b   => a-(a//b)*b
            # -10%3 => -10-(-10//3)*3
            #       => -10-(-4)*3
            #       => -10+4*3
            #       => -10+12
            #       =>  2

-3.3333333333333335
-4
2


In [ ]:
a,b,c,d,e=10,3,2,5,3
print(a*b-d//c+a%e*b+c)  # 1.()
                         # 2.** ??
                         # 3. *,/,%,// L to R
                         # 4. +,- L TO R
# a*b-d//c+a%e*b+c
# 30-d//c+a%e*b+c
# 30-2+a%e*b+c
# 30-2+1*b+c
# 30-2+3+2
# 33


33


In [ ]:
a,b=2,3
print(a**a**b)


# 2. Relational Operator <br>
 < , > , <= , >= , ==
  (Equal) ,  != (Not equal)

---
Relational Operator: Used for comparision of data.<br>
At an instance, can write only one condition <br>


---


Operator precedence <br>

1. Arithmetic
2. Relational



In [ ]:
a,b,c=map(int,input().split())
print(f' {a} < {b} : {a<b}')
print(f' {b} > {c} : {b>c}')
print(f' {a} <= {c} : {a<=c}')
print(f' {b} >= {a} : {b>=a}')
print(f' {c} == {a} : {c==a}')
print(f' {a} != {b} : {a!=b}')

11 12 13
 11 < 12 : True
 12 > 13 : False
 11 <= 13 : True
 12 >= 11 : True
 13 == 11 : False
 11 != 12 : True


In [ ]:
a,b,c=map(int,input().split())# 11 12 13
print(f' {a}+1 < {b} : {a+1<b}')# False
print(f' {b} > {c}-1 : {b>c-1}')#  False
print(f' {a} <= {c}-2 : {a<=c-2}')# True
print(f' {b}-1 >= {a} : {b-1>=a}')# True
print(f' {c}+0 == {a} : {c+0==a}')# False
print(f' {a} != {b}+2 : {a!=b+2}')# True

11 12 13
 11+1 < 12 : False
 12 > 13-1 : False
 11 <= 13-2 : True
 12-1 >= 11 : True
 13+0 == 11 : False
 11 != 12+2 : True


In [ ]:
n=int(input())
print(n>=5 n<=20)


# 03.Logical Operator <br>
 and , or , not <br>

 and => All condition in the given expresion must be true
 <br>

 or => Atleast one condtion must be true in the given expresion


 not => Inverse (True->False) <br>
 (False->True)


---
1. At an instance, can check more than one condition.

or
<br>
Multiple conditions can be written in a single instance.

2. Short circuit Evaluation
---
Operator precedence <br>
1. Arithmetic
2. Relational
3. Logical



In [ ]:
a=int(input())
print(f' {a}>=5 and {a}<=20 : {a>=5 and a<=20}')
print(f' {a}>=5 or {a}<=20 : {a>=5 or a<=20}')
print(f' not({a}>=5 and {a}<=20) : {not(a>=5 and a<=20)}')

-7
 -7>=5 and -7<=20 : False
 -7>=5 or -7<=20 : True
 not(-7>=5 and -7<=20) : True


In [ ]:
print(2<3 and 20 and 30 and 40) # any non zero value in place of
                                # will be considered as True

print(2<3 and 20 and 30 and 0 and 50)

print(10 or 20 or 30 or 40 or 50)
            # or => will stop if it encounters True

print(10 and 20 or 50 or 60) # and or => last updated truth value
 #          0 or 60
 #       0 and 50
print(0 and 0  and 50 or 60)

print(10 or 0 and 23 and 45)# or and => first updated truth value


40
0
10
20
60
10


# **04.Bitwise**<br>
1. & , | , ^(XOR) , <<(Left shift) , >>(Right Shift), ~(Negation).

In [ ]:
'''
& => both bits must be true

  10  => 0000 0011
   3  => 0000 1010  &
     ----------------
    =>    0000 0010  => ? 2
'''
a,b=10,3
print(a&b)


2


In [ ]:
'''
| => Atleast one  bit must be true

  10  => 0000 0011
   3  => 0000 1010  |
     ----------------
    =>   ---------
'''
a,b=10,3
print(a|b)


# loop

**Q7. Print the following series** <br>
Test case 1: <br>
Input: 6 <br>
Output: 0 3 8 15 24 35

In [ ]:
n = int(input())
res = []
for i in range(n):
    res.append(i * i + i * 2)
print(*res)


6
0 3 8 15 24 35


**Q8. Print the following series** <br>
Test case 1: <br>
Input: 5 <br>
Output: 5 10 15 20 25

In [ ]:
n = int(input())
for i in range(1, n + 1):
    print(i * 5, end=" ")


5
5 10 15 20 25 

**Q9. Print the following series** <br>
Test case 1: <br>
Input: 7 <br>
Output: 7 5 8 6 9 7 10 <br><br>

Test case 2: <br>
Input: 5 <br>
Output: 5 3 6 4 7




In [ ]:
n = int(input())

pos = []
neg = []

for i in range(n):
    pos.append(n + i)

for i in range(n // 2):
    neg.append(n - 2 + i)

res = []
for i in range(n):
    if i % 2 == 0:
        res.append(pos[i // 2])
    else:
        res.append(neg[i // 2])

print(*res)


7
7 5 8 6 9 7 10


**Q10. Print the following series** <br>
Test case 1:<br>
Input: 5 <br>
Output: 1 8 27 64 125 216


In [ ]:
n = int(input())
for i in range(1, n + 2):
    print(i * i * i, end=" ")


5
1 8 27 64 125 216 

**Q11. Print the following series** <br>
Test case 1:<br>
Input: 6 <br>
Output: 1 3 6 10 15 21


In [ ]:
n = int(input())
s = 0
for i in range(1, n + 1):
    s += i
    print(s, end=" ")


6
1 3 6 10 15 21 

**Q12. Print the following series** <br>
Test case 1:<br>
Input: 5 <br>
Output: 1 4 10 20 35


In [ ]:
n = int(input())
s = 0
for i in range(1, n + 1):
    s += i * i
    print(s, end=" ")


5
1 5 14 30 55 

**Q13. Print the following series** <br>
Test case 1:<br>
Input: 8 <br>
Output: 0 1 1 2 3 5 8 13


In [ ]:
n = int(input())
a = 0
b = 1
for i in range(n):
    print(a, end=" ")
    a, b = b, a + b


8
0 1 1 2 3 5 8 13 

# Pattern Printing

In [ ]:
for j in range(1,6):#j=1,2,3
    for i in range(1,6):
        print(j,end=" ")
    print()

1 1 1 1 1 
2 2 2 2 2 
3 3 3 3 3 
4 4 4 4 4 
5 5 5 5 5 


**Q2) Print the following Pattern**<br>
Test case 1:<br>
Input : 5<br>
Output: <br>
1 2 3 4 5<br>
1 2 3 4 5<br>
1 2 3 4 5<br>
1 2 3 4 5<br>
1 2 3 4 5<br><br>

In [ ]:
n = int(input())
for i in range(n):
    for j in range(1, n + 1):
        print(j, end=" ")
    print()


5
1 2 3 4 5 
1 2 3 4 5 
1 2 3 4 5 
1 2 3 4 5 
1 2 3 4 5 


**Print the following Pattern**<br>
Test case 1:<br>
Input : 5<br>
Output: <br>
5 4 3 2 1<br>
5 4 3 2 1<br>
5 4 3 2 1<br>
5 4 3 2 1<br>
5 4 3 2 1<br>



In [ ]:
n = int(input())
for i in range(n):
    for j in range(n, 0, -1):
        print(j, end=" ")
    print()


5
5 4 3 2 1 
5 4 3 2 1 
5 4 3 2 1 
5 4 3 2 1 
5 4 3 2 1 


**Print the following Pattern**<br>
Test case 1:<br>
Input : 5<br>
Output: <br>
1 1 1 1 1 <br>
1 0 0 0 1 <br>
1 0 0 0 1 <br>
1 0 0 0 1 <br>



In [ ]:
n = int(input())
for i in range(n):
    for j in range(n):
        if i == 0 or i == n-1 or j == 0 or j == n-1:
            print(1, end=" ")
        else:
            print(0, end=" ")
    print()


5
1 1 1 1 1 
1 0 0 0 1 
1 0 0 0 1 
1 0 0 0 1 
1 1 1 1 1 


**Print the following Pattern**<br>
Test case 1:<br>
Input : 5<br>
Output: <br>
1 <br>
1 2 <br>
1 2 3 <br>
1 2 3 4 <br>
1 2 3 4 5<br><br>



In [ ]:
n = int(input())
for i in range(1, n + 1):
    for j in range(1, i + 1):
        print(j, end=" ")
    print()


5
1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


**Print the following Pattern**<br>
Test case 1:<br>
Input : 5<br>
Output: <br>
1 <br>
2 1 <br>
3 2 1 <br>
4 3 2 1 <br>
5 4 3 2 1<br>



In [ ]:
n = int(input())
for i in range(1, n + 1):
    for j in range(i, 0, -1):
        print(j, end=" ")
    print()

5
1 
2 1 
3 2 1 
4 3 2 1 
5 4 3 2 1 


# Indexing

In [ ]:
#    0   1   2       3     4   5    6
l=['one',23,"apple",45.56,2+3j,78,'hello']
#  -7    -6   -5    -4     -3   -2   -1

print(l)
l.insert(2,'mango')
print(l)
l.insert(4,[3,5,6])
print(l)


['one', 23, 'apple', 45.56, (2+3j), 78, 'hello']
['one', 23, 'mango', 'apple', 45.56, (2+3j), 78, 'hello']
['one', 23, 'mango', 'apple', [3, 5, 6], 45.56, (2+3j), 78, 'hello']


In [ ]:
#    0   1   2       3     4   5    6
l=['one',23,"apple",45.56,2+3j,78,'hello']
#  -7    -6   -5    -4     -3   -2   -1

print(l)
l.remove('hello')
print(l)
#l.remove(100) Error


['one', 23, 'apple', 45.56, (2+3j), 78, 'hello']
['one', 23, 'apple', 45.56, (2+3j), 78]


In [ ]:
#    0   1   2       3     4   5    6
l=['one',23,"apple",45.56,2+3j,78,'hello']
#  -7    -6   -5    -4     -3   -2   -1

print(l)
l.pop(3)
deleted_obj = l.pop(-2)
print(deleted_obj)
l.pop()
#l.pop(12) Error

In [ ]:
#    0   1   2       3     4   5    6
l=['one',23,"apple",45.56,2+3j,78,'hello']
#  -7    -6   -5    -4     -3   -2   -1

print(l)
l.clear()
print(l)


['one', 23, 'apple', 45.56, (2+3j), 78, 'hello']
[]


In [ ]:
#    0   1   2       3     4   5    6
l=['one',23,"apple",45.56,2+3j,78,'hello']
#  -7    -6   -5    -4     -3   -2   -1

print(l)
del l
#print(l) Error

In [ ]:
#index(self, value, start=0, stop=9223372036854775807, /)
#      Return first index of value.
#   0 1 2 3 4 5 6 7 8 9 10 11 12 13
a1=[1,2,2,3,4,1,2,3,3,32,1,5, 6, 3]
print(a1.index(2))# 1

print(a1.index(2,4,14))#6

print(a1.index(4,3))

#print(a1.index(2,-1,-14)) Error

#print(a1.index('hello')) Error


In [ ]:
a1=[1,2,2,3,4,1,2,3,3,32,1,5, 6, 3]
print(a1)
print(f' a1.count(3) ={a1.count(3)}')

[1, 2, 2, 3, 4, 1, 2, 3, 3, 32, 1, 5, 6, 3]
 a1.count(3) =4


In [ ]:
a=[1,2,3,'abc',12.234,2+3j]
b=a # the reference of the list will be initialized
b[2]='apple'
print(a,b)

c=a.copy() # creates a duplicate list
c[1]='cat'
print(f'a={a}, c={c}')


[1, 2, 'apple', 'abc', 12.234, (2+3j)] [1, 2, 'apple', 'abc', 12.234, (2+3j)]
a=[1, 2, 'apple', 'abc', 12.234, (2+3j)], c=[1, 'cat', 'apple', 'abc', 12.234, (2+3j)]


In [ ]:
l=[1,2,2,3,4,51,2,123]
print(l)
l.reverse()
print(l)

[1, 2, 2, 3, 4, 51, 2, 123]
[123, 2, 51, 4, 3, 2, 2, 1]


In [ ]:
l=[1,8,3,5,2,6,3,6,2,4]
print(l)
l.sort()#ascending
print(l)
l.sort(reverse= True)#descending
print(l)

[1, 8, 3, 5, 2, 6, 3, 6, 2, 4]
[1, 2, 2, 3, 3, 4, 5, 6, 6, 8]
[8, 6, 6, 5, 4, 3, 3, 2, 2, 1]


In [ ]:
a=['hello','hi','one','three']
print(a)
a.sort(key=len) # sorts based on length
                # of the object
print(a)
a.sort(key=len,reverse=True)
print(a)

['hello', 'hi', 'one', 'three']
['hi', 'one', 'hello', 'three']
['hello', 'three', 'one', 'hi']


In [ ]:
l=[1,2,3,2,1,4,5]
print(l)
print(f' sum(l) :{sum(l)}')
print(f' max(l) :{max(l)}')
print(f' min(l) :{min(l)}')
print(f' len(l) :{len(l)}',end='\n\n')

#all() => checks all the objects in list
#      are non zero values
# l=[1,2,3]
#all(l) or all([1,2,3])
print(f' all([1,2,3]) :{all([1,2,3])}')
print(f' all([1,0,2]) :{all([1,0,2])}',end='\n\n')

#any() => check for atleast one, non-zero value
print(f' any([1,2,3]) :{any([1,2,3])}')
print(f' any([1,0,2]) :{any([1,0,2])}')
print(f' any([0,0,0]) :{any([0,0,0])}')


[1, 2, 3, 2, 1, 4, 5]
 sum(l) :18
 max(l) :5
 min(l) :1
 len(l) :7

 all([1,2,3]) :True
 all([1,0,2]) :False

 any([1,2,3]) :True
 any([1,0,2]) :True
 any([0,0,0]) :False


In [ ]:
#  0 1 2 3     4      5  6     7
l=[1,2,3,4,[10,20,30],23,45,[5,6,7]]
#            0  1  2         0 1 2

print(len(l))
print(f' l[3]={ l[3] }, l[5]={ l[5] }')
print(f' l[4]={ l[4] }, l[7]={ l[7] }')
print(f' l[4][1]= {l[4][1]}, l[7][2]={ l[7][2] } ')
print(l[4][::-1])
print(l)
l[4].reverse()
print(l)

8
 l[3]=4, l[5]=23
 l[4]=[10, 20, 30], l[7]=[5, 6, 7]
 l[4][1]= 20, l[7][2]=7 
[30, 20, 10]
[1, 2, 3, 4, [10, 20, 30], 23, 45, [5, 6, 7]]
[1, 2, 3, 4, [30, 20, 10], 23, 45, [5, 6, 7]]


In [ ]:
#  0 1       2            3
#       0  1  2  3    0   1  2  3  4
l=[1,2,[10,20,30,40],[11,12,13,14,15]]
#       -4 -3 -2 -1   -5 -4 -3 -2  -1
# -4,-3      -2           -1
print(l[2])
# l[][start:]
print(l[2][2:], l[2][1:])

# l[][:stop]
print(l[2][:3],l[2][:-1])#10,20,30
print(l[3][:3],l[3][:-2])

#l[][::step]
print(l[2][::-1])
print(l[-1][::-2])

print(l[-2][1:3:-1])


[10, 20, 30, 40]
[30, 40]


In [ ]:
l='hello' if 10>9 else "hi"
print(l)

hello


In [ ]:
for i in range(10):
    print(i,end=' ')

0 1 2 3 4 5 6 7 8 9 

In [ ]:
l=[i for i in range(10)]
print(l)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [ ]:
l= [i for i in range(1,11,2)]
print(l)

[1, 3, 5, 7, 9]


In [ ]:
l= [i for i in range(0,11,2)]
print(l)

[0, 2, 4, 6, 8, 10]


In [ ]:
a=[10,23,56,12,87,21,65,89,100]
l=[i for i in a]
print(l)

[10, 23, 56, 12, 87, 21, 65, 89, 100]


In [ ]:
a=[10,23,56,12,87,21,65,89,100]
L2= [i for i in a if i%2==0]
print(L2)

[10, 56, 12, 100]


In [ ]:
# 0 3 8 15 24 35
print(*[i**2-1 for i in range(1,int(input())+1)])


6
0 3 8 15 24 35


In [ ]:
l=[]
for i in range(5):
    l.append( int(input()))

print(l)

5
1
2
3
4
[5, 1, 2, 3, 4]


In [ ]:
l=[int(input()) for i in range(int(input()))]
print(l)

2
5
6
[5, 6]


In [ ]:
outer=[]
for i in range(3):
    outer.append([])
print(outer)

[[], [], []]


In [ ]:
outer=[]
for i in range(3):
    l=list(map(int,input().split()))
    outer.append(l)
print(outer)

5
6
7
[[5], [6], [7]]


In [ ]:
t='Enter number of nested list to insert :'
l= [ list(map(int,input().split()))
        for i in range(int(input(t)))]
print(l)

Enter number of nested list to insert :3
1 2 3 4
4 5 6 
7 8 9 12 34 67 78 890
[[1, 2, 3, 4], [4, 5, 6], [7, 8, 9, 12, 34, 67, 78, 890]]


In [ ]:
#                       i
l=[10,1,1,2,3,1,4,1,3,1,23]
unique=[]
for i in l:
    if i not in unique:
        unique.append(i)# 10 1 2 3 4 23

print(unique)



[10, 1, 2, 3, 4, 23]


#Tuples

In [ ]:
l = []  # list
t = ()
print(t, type(t))

t = tuple()  # create an empty tuple
print(t, type(t))


() <class 'tuple'>
() <class 'tuple'>


In [ ]:
t=(1,2,"niet",'batch4',12+0j,21.43)
print(t)

(1, 2, 'niet', 'batch4', (12+0j), 21.43)


In [ ]:
t=tuple([1,2,3,4])
print(t)

(1, 2, 3, 4)


In [ ]:
t=tuple(input().split())
print(t)

1 2 3 4 5 6
('1', '2', '3', '4', '5', '6')


In [ ]:
t=tuple(map(int,input().split()))
print(t)

1 2 3 4 5 6 8 9 0
(1, 2, 3, 4, 5, 6, 8, 9, 0)


In [ ]:
t=(1,2,"niet",'batch4',12+0j,21.43)
print(t)

for i in t:
    print(i,end='  ')

print(*t)

(1, 2, 'niet', 'batch4', (12+0j), 21.43)
1  2  niet  batch4  (12+0j)  21.43  1 2 niet batch4 (12+0j) 21.43


In [ ]:
#  0  1   2      3        4    5
t=(1, 2,"niet",'batch4',12+0j,21.43)
#  -6 -5  -4      -3     -2    -1

print(f' t[2] = {t[2]}, t[-4]={ t[-4]}')
print(f' t[4] = {t[4]}, t[-2]={ t[-2]}')

# t[0]="new object" Error
print(t)



 t[2] = niet, t[-4]=niet
 t[4] = (12+0j), t[-2]=(12+0j)
(1, 2, 'niet', 'batch4', (12+0j), 21.43)


In [ ]:
#  0  1   2      3        4    5
t=(1, 2,"niet",'batch4',12+0j,21.43)
#  -6 -5  -4      -3     -2    -1

print(f' t[1:5:1] ={t[1:5:1]}')
print(f' t[1:5:-1]= {t[1:5:-1]}')
print(f' t[5:1:-1]= {t[5:1:-1]}')


(2, 'niet', 'batch4', (12+0j))


In [ ]:
#  0  1   2      3        4    5
t=(1, 2,"niet",'batch4',12+0j,21.43)
#  -6 -5  -4      -3     -2    -1

n= t[:3]+t[3::-1]
print(n)

n= t[:3:-1]+t[:3:1]
print(n)

n=t[::-2]+t[::2]
print(n)



(1, 2, 'niet', 'batch4', 'niet', 2, 1)
(21.43, (12+0j), 1, 2, 'niet')
(21.43, 'batch4', 2, 1, 'niet', (12+0j))


In [ ]:
t=(1,2,3,(34,56,78,90),"hello")
print(t[3][1],t[-2][-3])
print(t[-2][1],t[3][-3])
print(t[3][::-1])
print(t[-2][::-1])

In [ ]:
#  0  1  2  3  4  5  6  7  8  9 10 11  12 13 14
t=(1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5)
print(t)
print(t.count(2))
print(t.count(100))
print(t.index(3))
print(t.index(3,3,14))
del t # deletes the tuple

In [ ]:
t = tuple([i for i in range(10)])
print(t)

(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)


In [ ]:
print(tuple([2*i for i in range(1,int(input())+1)]))

10
(2, 4, 6, 8, 10, 12, 14, 16, 18, 20)


In [ ]:
t1=(1,2,3)
t2=((4,5,6,7),)
print(len(t2))
t2=((4,5,6,7))
print(len(t2))
t4=((4,5,6,7),78)
print(len(t4))
t3=t1+t2
print(len(t3))

1
4
2
7


In [ ]:
t=tuple(input().split())
key = input()
if key in t:
    print(t[:t.index(key)]+t[t.index(key)+1:])
else:
    print(t)

1 2 3 4 5
1
('2', '3', '4', '5')


In [ ]:

t=(1,2,3,'hello',2+3j,(1,2,3),4,5)
c=0
for i in t:
    if isinstance(i,tuple):
        break
    c+=1 #5

print(c)# 5


5


# Sets

In [ ]:
s={}
print(s,type(s))

{} <class 'dict'>


In [ ]:
s={1}
print(s,type(s))

{1} <class 'set'>


In [ ]:
s=set()
print(s,type(s))

set() <class 'set'>


In [ ]:
s={1,2,8.0,2,1,3,5,4,2,7,2,9,5}
print(s)

{1, 2, 3, 4, 5, 7, 8.0, 9}


In [ ]:
s={1,2,8.0,2,1,3,5,4,2,7,2,9,5}
for i in s:
    print(i,end=' ')

1 2 3 4 5 7 8.0 9 

In [ ]:
help(set)

Help on class set in module builtins:

class set(object)
 |  set() -> new empty set object
 |  set(iterable) -> new set object
 |
 |  Build an unordered collection of unique elements.
 |
 |  Methods defined here:
 |
 |  __and__(self, value, /)
 |      Return self&value.
 |
 |  __contains__(...)
 |      x.__contains__(y) <==> y in x.
 |
 |  __eq__(self, value, /)
 |      Return self==value.
 |
 |  __ge__(self, value, /)
 |      Return self>=value.
 |
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |
 |  __gt__(self, value, /)
 |      Return self>value.
 |
 |  __iand__(self, value, /)
 |      Return self&=value.
 |
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |
 |  __ior__(self, value, /)
 |      Return self|=value.
 |
 |  __isub__(self, value, /)
 |      Return self-=value.
 |
 |  __iter__(self, /)
 |      Implement iter(self).
 |
 |  __ixor__(self, value, /)
 |      Return self^=value.
 |
 |  __l

In [ ]:
s=set(map(int,input().split()))
print(s)

12 45 12 56 879 35 312 89 312 879
{312, 35, 12, 45, 879, 56, 89}


In [ ]:
#s.add(10,12) Error
#s.add({10,11,12}) Error
# add takes only one arg

s={312, 35, 12, 45, 879, 56, 89}
s.add(23)
print(s)

{35, 23, 312, 89, 56, 12, 45, 879}


In [ ]:
#s.update(12) Error
# update accepts only iterables

s={312, 35, 12, 45, 879, 56, 89}
s.update({12,34,45,67})
print(s)
s.update([11,12,13,14,15])
print(s)

In [ ]:
s={312, 35, 12, 45, 879, 56, 89}
a=s.copy()
print(a)

{35, 312, 89, 56, 12, 45, 879}


In [ ]:
s={312, 35, 12, 45, 879, 56, 89}
print(s.pop())# random object from set will be deleted
print(s)
s.remove(45)
print(s)
#s.remove(500) Error
s.discard(312)
print(s)
s.discard(500)
s.clear()
print(s)
del s

35
{312, 89, 56, 12, 45, 879}
{312, 89, 56, 12, 879}
{89, 56, 12, 879}
set()


In [ ]:
s1={10,11,12,13,14,15}
s2={9,8, 7, 6, 5, 4}
s1.union(s2)
print(f's1={s1}, s2={s2}')
s3=s1.union(s2)
print(f's3={s3}')
s4=s1|s2
print(f' s4={s4}')

s1={10, 11, 12, 13, 14, 15}, s2={4, 5, 6, 7, 8, 9}
s3={4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}
 s4={4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}


In [ ]:
s1={10,11,12,13,4,5}
s2={9,8, 7, 6, 5, 4}
s1.intersection(s2)
print(f's1={s1} s2={s2}')
s3=s1.intersection(s2)
print(f' s3={s3}')
s3= s1 & s2
print(f' s3={s3}')

s1={4, 5, 10, 11, 12, 13} s2={4, 5, 6, 7, 8, 9}
 s3={4, 5}
 s3={4, 5}


In [ ]:
s1={10,11,12,13,4,5}
s2={9,8, 7, 6, 5, 4}
s1.intersection_update(s2)
print(f's1={s1} s2={s2}')

s1={4, 5} s2={4, 5, 6, 7, 8, 9}


In [ ]:
s1={10,11,12,13,4,5}
s2={9,8, 7, 6, 5, 4}
s1&= s2
print(f's1={s1} s2={s2}')

s1={4, 5} s2={4, 5, 6, 7, 8, 9}


In [ ]:
s1={10,11,12,13,4,5}
s2={9,8, 7, 6, 5, 4}
s3=s1.difference(s2)
print(f's1={s1}, s2={s2}')
print(f's3={s3}')
s3= s1-s2
print(f's1={s1}, s2={s2}')
print(f's3={s3}')

s1={4, 5, 10, 11, 12, 13}, s2={4, 5, 6, 7, 8, 9}
s3={10, 11, 12, 13}
s1={4, 5, 10, 11, 12, 13}, s2={4, 5, 6, 7, 8, 9}
s3={10, 11, 12, 13}


In [ ]:
s1={10,11,12,13,4,5}
s2={9,8, 7, 6, 5, 4}
print(f's1={s1}, s2={s2}')
s1.difference_update(s2)
print(f's1={s1}, s2={s2}')
s1 = s1-s2 # s1-=s2
print(f's1={s1}, s2={s2}')

s1={4, 5, 10, 11, 12, 13}, s2={4, 5, 6, 7, 8, 9}
s1={10, 11, 12, 13}, s2={4, 5, 6, 7, 8, 9}
s1={10, 11, 12, 13}, s2={4, 5, 6, 7, 8, 9}


In [ ]:
s1={10,11,12,13,4,5}
s2={9,8, 7, 6, 5, 4}
print(f's1={s1}, s2={s2}')
s3= s1.symmetric_difference(s2)
print(f' s3={s3}')
s3=s1^s2
print(f' s3={s3}')

s1={4, 5, 10, 11, 12, 13}, s2={4, 5, 6, 7, 8, 9}
 s3={6, 7, 8, 9, 10, 11, 12, 13}
 s3={6, 7, 8, 9, 10, 11, 12, 13}


In [ ]:
s1={10,11,12,13,4,5}
s2={9,8, 7, 6, 5, 4}
print(f's1={s1}, s2={s2}')
s1.symmetric_difference_update(s2)
print(f' s1={s1}')
s1 =s1^s2

s1={4, 5, 10, 11, 12, 13}, s2={4, 5, 6, 7, 8, 9}
 s1={6, 7, 8, 9, 10, 11, 12, 13}


In [ ]:
s1={1,2,3,4,5,6,7,8}
s2={3,4,5}
print(s2.issubset(s1))
s2={3,4,5,10,11}
print(s2.issubset(s1))

True
False


In [ ]:
s1={1,2,3,4}
s2={1,2,3,4}
print(s2.issubset(s1))
print(s2<=s1)

print(s2<s1)#Proper subset
# the no of object in s1 should be greater than s2

s1={1,2,3,4,5}
s2={1,2,3,4}
print(s2<s1)

True
True
False
True


In [ ]:
s1={1,2,3,4}
s2={1,2,5,56,6}
print(s2.isdisjoint(s1))
s1={1,2,3,4}
s2={5,6,7,8}
print(s2.isdisjoint(s1))

False
True


In [ ]:
s1 ={1,2,3,4,5,6,7,8}
s2={3,4,5,6,7}
print(s1.issuperset(s2))

True


In [ ]:
s1={1,2,3,4}# set()=> mutable
print(s1)
s1.add(-3)
print(s1)

{1, 2, 3, 4}
{1, 2, 3, 4, -3}


In [ ]:
s1=frozenset({1,2,3,4})
print(s1)# frozenset()=> immutable
#s1.add(-3) Error

frozenset({1, 2, 3, 4})


In [ ]:
s= {i*2 for i in range(int(input())) }
print(s)

5
{0, 2, 4, 6, 8}


In [ ]:
print({i for i in [1,2,1,2,3,1,4,1,4]})

{1, 2, 3, 4}


In [ ]:
# s1={ 1,{2,3}} Error
s1 ={1, frozenset({1,2,3})}
print(s1)

{frozenset({1, 2, 3}), 1}


# Dictionary

In [ ]:
d={}
print(d,type(d))
d=dict()
print(d,type(d))

{} <class 'dict'>
{} <class 'dict'>


In [ ]:
d={ 1:'one', 'mango':60, 'k':56.45, \
                        12.34:'asd' }
print(d)

{1: 'one', 'mango': 60, 'k': 56.45, 12.34: 'asd'}


In [ ]:
d={'apple':100, 'mango':40, 'coconut':30}
print(d)
print(d['mango'],d['apple'],d['coconut'])
print(" d[mango]=",d['mango'])
print(" d[apple]=",d['apple'])
print(" d[coconut]=",d['coconut'])

{'apple': 100, 'mango': 40, 'coconut': 30}
40 100 30
 d[mango]= 40
 d[apple]= 100
 d[coconut]= 30


In [ ]:
d={'apple':100, 'mango':40, 'coconut':30}
d['orange']=70
for i in d:
    print(f'i={i},d[i]={d[i]}')
    print(f'd[{i}] ={d[i]}',end=' \n\n ')

In [ ]:
d={}
n=int(input('Enter No.of key Value Pairs :'))
for i in range(n):
    key=input('Enter key:')
    value= input('Enter value:')
    d[key]=value

print(d)


Enter No.of key Value Pairs :3
Enter key:1
Enter value:10
Enter key:2
Enter value:20
Enter key:3
Enter value:30
{'1': 10, '2': 20, '3': 30}


In [ ]:
d={1:'one', 2:'Two',3:'Four',5:'five',1:'12'}
print(d)
d[3]='three'
print(d)
d[6]='six'
print(d)
d[1]="new"
print(d)
#duplicate keys are not allowed in dict
#but 'values'=> can be re-initialized

In [ ]:
d={1:{1,2,3,4,5}, 'key2':['a','b','c'], 3:(3,4,5)}
print(d)

{1: {1, 2, 3, 4, 5}, 'key2': ['a', 'b', 'c'], 3: (3, 4, 5)}


In [ ]:
d={'key1':10, 'key2':{1:'one', 2:'two',3:'three'}}
print(d)
print("d['key1']=",d['key1'])
print("d['key2']=",d['key2'])

# Accessing nested dict
print("d['key2'][1]=",d['key2'][1])
print("d['key2'][2]=",d['key2'][2])
print("d['key2'][3]=",d['key2'][3])
# adding a new key value pair to nested dict
d['key2'][4]='four'
print(d['key2'])

{'key1': 10, 'key2': {1: 'one', 2: 'two', 3: 'three'}}
d['key1']= 10
d['key2']= {1: 'one', 2: 'two', 3: 'three'}
d['key2'][1]= one
d['key2'][2]= two
d['key2'][3]= three
{1: 'one', 2: 'two', 3: 'three', 4: 'four'}


In [ ]:
d={'a':'app', 'b':123, 'c':2+3j, 'd':True, 'q':'abc'}
print(d)
print("d.get('a')=",d.get('a'))
print("d.get('b')=",d.get('b'))
print("d.get('c')=",d.get('c')) # d['c']

print("d.keys()= ",d.keys())
print("d.values()= ", d.values())
print("d.items()= ",d.items())

{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}
d.get('a')= app
d.get('b')= 123
d.get('c')= (2+3j)
d.keys()=  dict_keys(['a', 'b', 'c', 'd', 'q'])
d.values()=  dict_values(['app', 123, (2+3j), True, 'abc'])
d.items()=  dict_items([('a', 'app'), ('b', 123), ('c', (2+3j)), ('d', True), ('q', 'abc')])


In [ ]:
d={'a':'app', 'b':123, 'c':2+3j, 'd':True, 'q':'abc'}
print(d)
new_dict ={1:'one', 2:'two', 3:'three'}

d.update({'new_key':10})
print(d)
d.update(new_dict)
print(d)
d.update({'a':'apple'})
print(d)

{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}
{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc', 'new_key': 10}
{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc', 'new_key': 10, 1: 'one', 2: 'two', 3: 'three'}
{'a': 'apple', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc', 'new_key': 10, 1: 'one', 2: 'two', 3: 'three'}


In [ ]:
d={'a':'app', 'b':123, 'c':2+3j, 'd':True, 'q':'abc'}
print(d)
b=d
b=d.copy()
print(b)


{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}
{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}


In [ ]:
d={'a':'app', 'b':123, 'c':2+3j, 'd':True, 'q':'abc'}
print(d)
d.popitem()
print(d)

{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}
{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True}


In [ ]:
d={'a':'app', 'b':123, 'c':2+3j, 'd':True, 'q':'abc'}
print(d)
d.pop('a')
print(d)

{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}
{'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}


In [ ]:
d={'a':'app', 'b':123, 'c':2+3j, 'd':True, 'q':'abc'}
print(d)
d.clear()
print(d)
del d

{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}
{}


In [ ]:
d={'a':'app', 'b':123, 'c':2+3j, 'd':True, 'q':'abc'}
print(d)
k={}
k=d.fromkeys(d)
print(k)
k=dict.fromkeys(d,10)
print(k)
k=dict.fromkeys([1,2,3,4,5],'batch4')
print(k)



{'a': 'app', 'b': 123, 'c': (2+3j), 'd': True, 'q': 'abc'}
{'a': None, 'b': None, 'c': None, 'd': None, 'q': None}
{'a': 10, 'b': 10, 'c': 10, 'd': 10, 'q': 10}
{1: 'batch4', 2: 'batch4', 3: 'batch4', 4: 'batch4', 5: 'batch4'}


In [ ]:
d={1:10, 2:20, 3:30, 4:40}
print(d)
print(d.setdefault(2))
print(d.setdefault(5))
print(d)
print(d.setdefault(6,45))
print(d)
print(d.setdefault(2,35))
print(d)


{1: 10, 2: 20, 3: 30, 4: 40}
20
None
{1: 10, 2: 20, 3: 30, 4: 40, 5: None}
45
{1: 10, 2: 20, 3: 30, 4: 40, 5: None, 6: 45}
20
{1: 10, 2: 20, 3: 30, 4: 40, 5: None, 6: 45}


In [ ]:
d={1:10, 2:20, 3:30, 4:40}
d2= {i:d[i] for i in d}
print(d2)

{1: 10, 2: 20, 3: 30, 4: 40}


In [ ]:
d={1:10, 2:20, 3:30, 4:40}
d2 = {k:v for k,v in d.items()}
print(d2)

{1: 10, 2: 20, 3: 30, 4: 40}


In [ ]:
d= {i:i**2 for i in range(10)}
print(d)

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}


In [ ]:
help(dict)

#**string**

In [ ]:
s='batch 4'
print(s)
s="batch 4"
print(s)

s='batch 4 \
   new line'
print(s)
s="batch 4  \
   new line"
print(s)

s='''batch 4
     new line'''
print(s)
s="""batch 4
    newline"""
print(s)

batch 4
batch 4
batch 4    new line
batch 4     new line
batch 4
     new line
batch 4
    newline


In [ ]:
#  012345678910
s='hello world'
#-11.........-1

print(f' s[0]={s[0]}, s[-11]={s[-11]}')
print(f' s[3]={s[3]}, s[-8]={s[-8]}')
print(f' s[5]={s[5]}, s[-6]={s[-6]}')
print(f' s[10]={s[10]}, s[-1]={s[-1]}')

 s[0]=h, s[-11]=h
 s[3]=l, s[-8]=l
 s[5]= , s[-6]= 
 s[10]=d, s[-1]=d


In [ ]:
#  012345678910
s='hello world' # immutable
#-11.........-1
print(s)
#s[0]='c' Error
print(s)

hello world
hello world


In [ ]:
s='hello world'
print(s)

for i in range(len(s)):
    print(s[i],end='')

print()

for i in s:
    print(i,end='')


hello world
hello world
hello world

In [ ]:
#  01234567891011
s='ALL THE BEST'
print(s[0:])
print(s[4:])
print(s[8:],end='\n\n')

print(s[:5])
print(s[:9],end='\n\n')

print(s[::1])
print(s[::2])
print(s[::3])


ALL THE BEST
THE BEST
BEST

ALL T
ALL THE B

ALL THE BEST
ALTEBS
A EE


In [ ]:
s='mango'
print(s[::-1])
print(s[-3::-1]+s[2:])

ognam
namngo


In [ ]:
s=input()
print(s)

my name is Anurag Verma
my name is Anurag Verma


In [ ]:
import string
help(string)

Help on module string:

NAME
    string - A collection of string constants.

MODULE REFERENCE
    https://docs.python.org/3.12/library/string.html

    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    Public module variables:

    whitespace -- a string containing all ASCII whitespace
    ascii_lowercase -- a string containing all ASCII lowercase letters
    ascii_uppercase -- a string containing all ASCII uppercase letters
    ascii_letters -- a string containing all ASCII letters
    digits -- a string containing all ASCII decimal digits
    hexdigits -- a string containing all ASCII hexadecimal digits
    octdigits -- a string containing all ASCII octal digits
    punctuation -- a string containing all A

In [ ]:
'''
 ascii_letters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
    ascii_lowercase = 'abcdefghijklmnopqrstuvwxyz'
    ascii_uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    digits = '0123456789'
    hexdigits = '0123456789abcdefABCDEF'
    octdigits = '01234567'
    printable = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTU...
    punctuation = '!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'
    whitespace = ' \t\n\r\x0b\x0c'

'''

'\n ascii_letters = \'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ\'\n    ascii_lowercase = \'abcdefghijklmnopqrstuvwxyz\'\n    ascii_uppercase = \'ABCDEFGHIJKLMNOPQRSTUVWXYZ\'\n    digits = \'0123456789\'\n    hexdigits = \'0123456789abcdefABCDEF\'\n    octdigits = \'01234567\'\n    printable = \'0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTU...\n    punctuation = \'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~\'\n    whitespace = \' \t\n\r\x0b\x0c\'\n\n'

In [ ]:
import string
print(string.ascii_letters)
print(string.ascii_lowercase)
print(string.ascii_uppercase)
print(string.digits)
print(string.printable)
print(string.punctuation)
print(string.whitespace)
print(string.hexdigits)
print(string.octdigits)

abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
abcdefghijklmnopqrstuvwxyz
ABCDEFGHIJKLMNOPQRSTUVWXYZ
0123456789
0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ 	

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
 	

0123456789abcdefABCDEF
01234567


In [ ]:
s='all the Best'
print(s.capitalize())
print(s.title())
s='aLL thE bEst'
print(s.casefold())
print(s.upper())
print(s.lower())
print(s.swapcase())

All the best
All The Best
all the best
ALL THE BEST
all the best
All THe BeST


In [ ]:
a=10;b=20;c=30
print(a,b,c)
print('a=',a,', b=',b,' ,c=',c)
print(" a= %d, c=%d, d=%d" % (a,b,c))

10 20 30
a= 10 , b= 20  ,c= 30
 a= 10, c=20, d=30


In [ ]:
a=10;b=20;c=30
print("a={}, b={}, a={} ".format(a,b,a))

a=10, b=20, c=30 


In [ ]:
a=10;b=20;c=30
                          #        0 1
print("a={0}, b={1}, a={0} ".format(a,b))

a=10, b=20, a=10 


In [ ]:
firstIdInt=10;secondIdInt=20
print('b={b},a={a}'.format(a=firstIdInt, \
                           b=secondIdInt))

b=20,a=10


In [ ]:
a=11234
print("{:3.2%}".format(a))
print("{:.3f}".format(a))
print("{:.2f}".format(a))
print("{:.7f}".format(a))

1123400.00%
11234.000
11234.00
11234.0000000


In [ ]:
a="Vegetables"
print(a.endswith('s'))
print(a.endswith('tables'))
print(a.endswith('tables '))
print(a.startswith('veg'))
print(a.startswith('Veg'))
a.replace('e','i')
a.replace('e','abc')

True
True
False
False
True


'Vabcgabctablabcs'

In [ ]:
a='arun all the best tarun'
print(a.partition('t'))
print(a.index('t'))
print(a.rindex('t'))
print(a.count('t'))
print(a.find('arun'))
print(a.rfind('arun'))

('arun all ', 't', 'he best tarun')
9
18
3
0
19


In [ ]:
a='123'
print(a.isdigit())
a='123a'
print(a.isalnum())
a='ads'
print(a.islower())
print(a.isupper())
a='12345'
print(a.isnumeric())
a='12.345'
print(a.isnumeric())
print(a.isdecimal())
a='batch4'
print(a.isprintable())
a='batch4 \n'
print(a.isprintable())
a='\t'
print(a.isspace())



True
True
True
False
True
False
False
True
False
True


In [ ]:
s="hello"
print(len(s))
print(max(s))
print(min(s))

5
o
e


Test case 1:<br>
input: All the best<br>
output: <br>
ALL <br>
        THE<br>
        BEST<br>

In [ ]:
s=input().split()
for i in s:
    print(i.upper())

All the best Anurag
ALL
THE
BEST
ANURAG
